addons/controllers/csi/vspherecsiconfig_controller.go

```diff
@@ -0,0 +1,173 @@
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *VSphereCSIConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("VSphereCSIConfig", req.NamespacedName)
+
+	vcsiConfig := &csiv1alpha1.VSphereCSIConfig{}
+	if err := r.Get(ctx, req.NamespacedName, vcsiConfig); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("VSphereCSIConfig resource not found")
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Unable to fetch VSphereCSIConfig resource")
+		return ctrl.Result{}, err
+	}
+
+	cluster, err := r.getOwnerCluster(ctx, vcsiConfig)
+	if cluster == nil {
+		return ctrl.Result{RequeueAfter: 20 * time.Second}, err // retry until corresponding cluster is found
+	}
+
+	return r.reconcileVSphereCSIConfig(ctx, vcsiConfig, cluster)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *VSphereCSIConfigReconciler) SetupWithManager(_ context.Context, mgr ctrl.Manager,
+	options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1alpha1.VSphereCSIConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+func (r *VSphereCSIConfigReconciler) reconcileVSphereCSIConfig(ctx context.Context,
+	csiCfg *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (result ctrl.Result, retErr error) {
+
+	logger := log.FromContext(ctx)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(csiCfg, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if retErr != nil {
+			// don't modify VSphereCSIConfig if there is an error
+			return
+		}
+
+		if err := patchHelper.Patch(ctx, csiCfg); err != nil {
+			logger.Error(err, "Failed to patch VSphereCSIConfig")
+			retErr = err
+		}
+	}()
+
+	if !csiCfg.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil // deleted
+	}
+
+	if result, err = r.reconcileVSphereCSIConfigNormal(ctx, csiCfg, cluster); err != nil {
+		logger.Error(err, "Error reconciling VSphereCSIConfig")
+		return result, err
+	}
+
+	return result, nil
+}
+
+func (r *VSphereCSIConfigReconciler) reconcileVSphereCSIConfigNormal(ctx context.Context,
+	csiCfg *csiv1alpha1.VSphereCSIConfig,
+	cluster *clusterapiv1beta1.Cluster) (ctrl.Result, error) {
+
+	logger := log.FromContext(ctx)
+
+	ownerRef := metav1.OwnerReference{
+		APIVersion: clusterapiv1beta1.GroupVersion.String(),
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+
+	if !clusterapiutil.HasOwnerRef(csiCfg.OwnerReferences, ownerRef) {
+		csiCfg.OwnerReferences = clusterapiutil.EnsureOwnerRef(csiCfg.OwnerReferences, ownerRef)
+	}
+
+	addonName := ""
+	switch csiCfg.Spec.VSphereCSI.Mode {
+	case VSphereCSINonParavirtualMode:
+		addonName = constants.CSIAddonName
+	case VSphereCSIParavirtualMode:
+		addonName = constants.PVCSIAddonName
+	}
```

Comment:
question: do we have a validation against `csiCfg.Spec.VSphereCSI.Mode` ? Trying to understand what addonName we want to assign if it is a no-match against above two cases.